Aggregated contributions

mgsloan opened the pull request:
I wanted to make some changes to keynav, so I took a look at what forks existed.  I merged the changes that looked quality to me, and then added some of my own.  My changes:

* Add `toggle-start` command, a combination of `start` and `end` based on if keynav is active

* Fix a bug where commands are run if there prefixes match.  E.g. `c` runs a bunch of commands and doesn't report an error.

* Fix a bug where keys are not ungrabbed on clear, if there are multiple commands which begin with `start`

I'd offer to maintain, but I do not frequently write code in C, my preferred language is Haskell.

jordansissel commented on keynav.c:
```diff
@@ -1440,13 +1536,16 @@ void viewport_left() {
   e->state &= ~(LockMask | Mod2Mask);
 
+  /* Ignore different keyboard layouts (e.g. russian) */
+  e->state &= ~(1<<13);
```
Is there a constant definition for this?

mgsloan replied:
Good point! 
 I think it makes more sense to have a whitelist of modifiers rather than a blacklist, made that change here: https://github.com/mgsloan/keynav/commit/2128fd26b0901f0208cbffdd50bdd6a41495b290

jordansissel commented on Makefile:
```diff
@@ -1,35 +1,40 @@
-CFLAGS+=$(shell pkg-config --cflags cairo-xlib xinerama glib-2.0 xext x11 xtst 2> /dev/null || echo -I/usr/X11R6/include -I/usr/local/include)
-LDFLAGS+=$(shell pkg-config --libs cairo-xlib xinerama glib-2.0 xext x11 xtst 2> /dev/null || echo -L/usr/X11R6/lib -L/usr/local/lib -lX11 -lXtst -lXinerama -lXext -lglib)
-LDFLAGS+=$(shell pkg-config --libs glib-2.0)
+CFLAGS+=$(shell pkg-config --cflags cairo-xlib 2> /dev/null)
```
Why split these up?

mgsloan replied:
Not sure, that wasn't my change.  It's been a *long* time since I've done much makefiles and C, so not sure what makes sense here.  It works fine for me either way.

jordansissel commented on Makefile:
```diff
@@ -1,35 +1,40 @@
 keynav: LDFLAGS+=-Xlinker -rpath=/usr/local/lib
 keynav: keynav.o
-	$(CC) keynav.o -o $@ $(LDFLAGS) -lxdo; \
+	$(CC) keynav.o -o keynav $(CFLAGS) $(LDFLAGS) -lxdo
```
why change $@ to `keynav` ? `$@` is the target name, which is `keynav` in this case. I'm a little confused; can you help me understand?

mgsloan replied:
Also not my change, pinging @yjftsjthsd-g

yjftsjthsd-g replied:
My memory fails me, unfortunately. I may have changed it just because `$@` was unfamiliar to me; looking now, it appears to be reasonably portable (the current version of the Makefile only works in GNU make, and I'd like to fix that), so I agree that it should be reverted. The only other minor concern would be use in ex. the debug target, but building a differently named debug binary isn't exactly a bad thing.

jordansissel commented on keynav.c:
```diff
@@ -216,7 +217,7 @@ dispatch_t dispatch[] = {
   // Mouse activity
   "warp", cmd_warp,
-  "click", cmd_click,     
+  "click", cmd_click,
```
The trailing whitespace changes are making the diff harder to read for me, so I am likely to miss some things in review.

mgsloan replied:
Sorry about that, my emacs config automatically trims it.

xtaran replied:
Maybe put all traling whitespace changes into a single commit at the very end of the list of patches.

jordansissel commented on keynav.c:
```diff
@@ -225,7 +226,8 @@ dispatch_t dispatch[] = {
   "start", cmd_start,
-  "end", cmd_end, 
+  "end", cmd_end,
+  "toggle-start", cmd_toggle_start,
```
++ this is a nice feature.

mgsloan replied:
Thanks!

jordansissel commented on keynav.c:
```diff
@@ -443,11 +466,25 @@ void parse_config() {
+  char *user_config_file;
+
+  if (config_home &&
+      asprintf(&user_config_file, "%s/keynav/keynavrc", config_home) != -1) {
```
asprintf is not available on Solaris (by default), but I suspect this is probably not a problem.

+1 to supporting XDG config pathing

mgsloan replied:
Cool!

yjftsjthsd-g replied:
Thanks for the example of where asprintf will fail; I noted that it wasn't portable but I've only been testing on GNU and BSD systems. I've updated https://github.com/yjftsjthsd-g/keynav/issues/6 to note that it'll break on Solaris.

jordansissel commented on keynav.c:
```diff
@@ -1591,26 +1720,28 @@ void handle_commands(char *commands) {
         found = 1;
         dispatch[i].func(args);
-
+        break;
```
this is ok, probably easier to read if the for loop condition is changed `dispatch[i].command && !found` instead of breaking here. (should have the same effect, though)

mgsloan replied:
Good point!  Made that change in https://github.com/mgsloan/keynav/commit/0331365944f5f5fdabc3b3e441eb6df92259bdc7

mgsloan commented:
@jordansissel Hey!  Thanks for the review.  I will try to address this stuff next week, been busy.  Most of those things are in code that other's wrote.  My changes also got merged to https://github.com/yjftsjthsd-g/keynav

mgsloan commented:
@jordansissel Hey!  Sorry for the delay on this.  I've added a couple commits addressing comments.  Can't do much about the Makefile changes.  Perhaps merge and add a commit putting the Makefile back, if you feel strongly about it?

mgsloan commented:
Also added a couple fixes that have been bugging me for years, particularly the xrandr thing.  No longer will I have to kill and restart keynav when plugging in a display!  Being a bit more familiar with the codebase helps.

jordansissel commented:
Sweet! I will review and merge when I get time hopefully today. Thank you
for making key available better!

On Sat, Sep 23, 2017 at 10:47 PM Michael Sloan <notifications@github.com>
wrote:

> Also added a couple fixes that have been bugging me for years,
> particularly the xrandr thing. No longer will I have to kill and restart
> keynav when plugging in a display! Being a bit more familiar with the
> codebase helps.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jordansissel/keynav/pull/23#issuecomment-331688906>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6sGpriZux7XZ_7S_Czi2fBrK0P84ks5slez8gaJpZM4PDTJC>
> .
>

sojusnik commented:
@jordansissel when do you plan to finally merge those fixes? Can't wait to use the up to date and feature rich version of keynav!

mgsloan commented:
I believe that the most up-to-date fork of keynav is
https://github.com/yjftsjthsd-g/keynav  .  Well plus my bugfix PR
https://github.com/yjftsjthsd-g/keynav/pull/15

On Fri, Aug 10, 2018 at 3:06 PM, Axel Beckert <notifications@github.com>
wrote:

> *@xtaran* commented on this pull request.
> ------------------------------
>
> In keynav.c
> <https://github.com/jordansissel/keynav/pull/23#discussion_r209395393>:
>
> > @@ -216,7 +217,7 @@ dispatch_t dispatch[] = {
>
>    // Mouse activity
>    "warp", cmd_warp,
> -  "click", cmd_click,
> +  "click", cmd_click,
>
> Maybe put all traling whitespace changes into a single commit at the very
> end of the list of patches.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jordansissel/keynav/pull/23#discussion_r209395393>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABYKkHsPAR6dnfFmILNAPT3e1K4hM1mks5uPgPbgaJpZM4PDTJC>
> .
>

jordansissel commented:
I haven't had much time for this project lately, so I appreciate your patience. The code overall looks good. I'll test it quickly and merge if it compiles and runs. :)

jordansissel commented:
Alrighty I did a quick test and things seems to be working ok.

I think this is a backwards compatible change (Everything I tested worked with my existing keynavrc). Merging!

jordansissel commented:
Thanks everyone for the combined efforts on these patches.

mgsloan commented:
No worries, I know how it goes.  Thanks for making keynav, I use it all the time.  I got a logitech mx ergo wireless mouse 4 months ago and have never needed to recharge the batteries.  Hah!

jerkey commented on README.md:
```diff
@@ -22,3 +22,35 @@ somewhere in your path). You can also install (by default directly to `/usr`)
+AutoHotKey (Windows), and AppleScript (macOS). If you find something that works,
+let me know and I'll consider adding it to this list.
+
+Q: Can I use keynav to scroll?  
```
I'm sorry if this information is easy to find, but is there a way to do control-click from keynav?
I tried putting "ctrl+1 click 1" in my .keynavrc but it had no effect - no click happens when i ctrl+1
(i also tried with ctrl-q and there's still no effect)
I just want to be able to control-click on something...am i missing something?  thank you

jordansissel commented:
Hmm... off the top of my head, I don’t know if there’s an easy way to do
keyboard-mouse combos like control+click. For today, you’d probably need to
do: xdotool keydown ctrl — click 1 — keyup ctrl

I support adding a way to do this if there isn’t one already :)

On Tue, Oct 30, 2018 at 5:52 PM jerkey <notifications@github.com> wrote:

> *@jerkey* commented on this pull request.
> ------------------------------
>
> In README.md
> <https://github.com/jordansissel/keynav/pull/23#discussion_r229535327>:
>
> > +A: keynav should work on nearly any Unix-like that runs X11. It has been
> +confirmed to work on extremely varied GNU/Linux systems (incuding RPM-based,
> +Debian derivatives, musl-based systems, and Arch), and FreeBSD. If you get it to
> +run elsewhere, please let me know so I can add it to the list. If you try to run
> +it on another Unix-like and have trouble, please get in touch and I'll try to
> +help. If attempting to run elsewhere, note that we currently have a dependency
> +on GNU Make (gmake), and it hasn't been tested with many compilers yet.
> +
> +Q: Does it work on Android/Windows/Wayland/iOS/...?
> +A: Sadly, no; keynav is totally dependent on X11, and porting it to any other
> +graphical system would really be a clone/rewrite. Although I am aware of no
> +exact analogues on other systems, I suggest looking into Tasker (Android),
> +AutoHotKey (Windows), and AppleScript (macOS). If you find something that works,
> +let me know and I'll consider adding it to this list.
> +
> +Q: Can I use keynav to scroll?
>
> I'm sorry if this information is easy to find, but is there a way to do
> control-click from keynav?
> I tried putting "ctrl+1 click 1" in my .keynavrc but it had no effect - no
> click happens when i ctrl+1
> (i also tried with ctrl-q and there's still no effect)
> I just want to be able to control-click on something...am i missing
> something? thank you
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jordansissel/keynav/pull/23#pullrequestreview-170058538>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6pS4pKzoPXNYGqGd2STQrFqwS7uZks5uqPRWgaJpZM4PDTJC>
> .
>
